presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotSegmentStreamingPageSource.java

```diff
@@ -0,0 +1,135 @@
+import com.facebook.presto.spi.ConnectorSession;
+import org.apache.pinot.common.utils.DataTable;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.pinot.PinotErrorCode.PINOT_DATA_FETCH_EXCEPTION;
+import static com.facebook.presto.pinot.PinotErrorCode.PINOT_INVALID_SEGMENT_QUERY_GENERATED;
+import static com.facebook.presto.pinot.PinotErrorCode.PINOT_UNEXPECTED_RESPONSE;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class retrieves Pinot data from a Pinot client, and re-constructs the data into Presto Pages.
+ */
+
+public class PinotSegmentStreamingPageSource
+        extends PinotSegmentPageSource
+{
+    private final PinotStreamingQueryClient pinotStreamingQueryClient;
+    private Iterator<ServerResponse> serverResponseIterator;
+    private long completedPositions;
+
+    public PinotSegmentStreamingPageSource(
+            ConnectorSession session,
+            PinotConfig pinotConfig,
+            PinotStreamingQueryClient pinotStreamingQueryClient,
+            PinotSplit split,
+            List<PinotColumnHandle> columnHandles)
+    {
+        super(session, pinotConfig, null, split, columnHandles);
+        this.pinotStreamingQueryClient = requireNonNull(pinotStreamingQueryClient, "pinotStreamingQueryClient is null");
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return closed;
+    }
+
+    /**
+     * @return constructed page for pinot data.
+     */
+    @Override
+    public Page getNextPage()
+    {
+        if (closed) {
+            return null;
+        }
+        if (serverResponseIterator == null) {
+            serverResponseIterator = queryPinot(split);
+        }
+        ByteBuffer byteBuffer = null;
+        if (serverResponseIterator.hasNext()) {
+            long startTimeNanos = System.nanoTime();
+            ServerResponse serverResponse = serverResponseIterator.next();
+            readTimeNanos += System.nanoTime() - startTimeNanos;
+            switch (serverResponse.getResponseType()) {
+                case Constants.Response.ResponseType.DATA:
+                    estimatedMemoryUsageInBytes = serverResponse.getSerializedSize();
+                    // Store each dataTable which will later be constructed into Pages.
+                    try {
+                        byteBuffer = serverResponse.getPayloadReadOnlyByteBuffer();
+                        DataTable dataTable = serverResponse.getDataTable(byteBuffer);
+                        checkExceptions(dataTable, split, PinotSessionProperties.isMarkDataFetchExceptionsAsRetriable(session));
+                        currentDataTable = new PinotSegmentPageSource.PinotDataTableWithSize(dataTable, serverResponse.getSerializedSize());
+                    }
+                    catch (IOException e) {
+                        throw new PinotException(
```

Comment:
Will anything be leaked if we throw here ? I am not sure if the byte buffer is native or on heap ? Should we free the byteBuffer after this loop using a try-finally ?

Reply:
good catch, I will wrapper with try- finally